### DISCUSSION TITLE
Creating a subgrid

### DISCUSSION THREAD

**Discussion by mkbenneh** [ufs-community/ufs-srweather-app#1155] [Q&A] [answered]
I made a predefined subgrid over Florida using the settings below; I have also included it in the valid_param_vals.yaml file, however it still doesn't work for me and will like to understand how making a subgrid work.
_"
GRID_GEN_METHOD: ESGgrid
  ESGgrid_LON_CTR: -80.5. 
  ESGgrid_LAT_CTR:  26
  ESGgrid_DELX: 3000.0
  ESGgrid_DELY: 3000.0
  ESGgrid_NX: 1820
  ESGgrid_NY: 1092
  ESGgrid_PAZI: 0.0
  ESGgrid_WIDE_HALO_WIDTH: 6
  DT_ATMOS: 36
  LAYOUT_X: 28
  LAYOUT_Y: 28
  BLOCKSIZE: 28
  QUILTING: 
    WRTCMP_write_groups: 1
    WRTCMP_write_tasks_per_group: 4
    WRTCMP_output_grid: lambert_conformal
    WRTCMP_cen_lon: '{{ task_make_grid.ESGgrid_LON_CTR }}'
    WRTCMP_cen_lat: '{{ task_make_grid.ESGgrid_LAT_CTR }}'
    WRTCMP_lon_lwr_left: -81.5
    WRTCMP_lat_lwr_left: 25.5
    WRTCMP_stdlat1: '{{ task_make_grid.ESGgrid_LAT_CTR }}'
    WRTCMP_stdlat2: '{{ task_make_grid.ESGgrid_LAT_CTR }}'
    WRTCMP_nx: 198
    WRTCMP_ny: 138
    WRTCMP_dx: 3000.0
    WRTCMP_dy: 3000.0
"_

**Comment by MichaelLueken** [comment]
Hello @mkbenneh,

Thank you for reaching out with your question regarding how to create a subgrid in the SRW App!  In order to assist with debugging your issue, there is some additional information that is required:

1. What version of the SRW App are you using (i.e., `release/public-v2.2.0` or the latest `develop`)?
2. What is the error message that you are receiving and is the experiment workflow properly being generated?
3. If your experiment is failing, please provide the log files for the failed tasks.
4. If you are running the SRW App on a Tier-1 platform (i.e., Hera, Jet, etc.), please provide a path so that the changes can be checked.

Thanks again and I look forward to assisting you with this issue.

Mike Lueken

**Reply by mkbenneh** [reply]
Hello @MichaelLueken,

The version I am running on derecho is release/public-v2.2.0. Here is the path (/glade/work/mbenneh/expt_dirs/Florida_try/log) to my new try. It fails at the plot_allvars.

Michael Benneh

**Reply by MichaelLueken** [reply]
Thanks, @mkbenneh!

Looking in `/glade/work/mbenneh/expt_dirs/Florida_try/log/plot_allvars_2019061518.log`, I see:

`OSError: [Errno could not open %s] /glade/work/mbenneh/expt_dirs/Florida_try/2019061518//postprd/srw.t18z.prslev.f000.seus_12km.grib2`

In your `config.yaml` file, I see that you have the following set:

```
PREDEF_GRID_NAME: SEUS_12km

POST_OUTPUT_DOMAIN_NAME: srw_seus_12km
```

Further, in the `postprd` directory, I see the output in question as:

`srw.t18z.natlev.f000.srw_seus_12km.grib2`

In your original `config.yaml` file, please try generating a new experiment after changing:

`POST_OUTPUT_DOMAIN_NAME: srw_seus_12km`

to

`POST_OUTPUT_DOMAIN_NAME: seus_12km`

The script should be able to find the post files following this modification.  If you continue to encounter issues, please let me know.

Mike Lueken

**Reply by mkbenneh** [reply]
Hello @MichaelLueken,

Works fine now, thank you. I think I would have to manipulate the grid more to get what I want. Also, in case I want to include more variables to plot, where do I look? And about the verification task in the config.yaml file, I believe I would have to include a path to a subdirectory with observation data, or is there some sort of default directory in decrecho for this task?

Michael Benneh

**Reply by MichaelLueken** [reply]
I'm glad that your experiment is working now, @mkbenneh!

With respect to the inclusion of more variables to plot, the current variables are read from the `/glade/work/mbenneh/ufs-srweather-app/parm/upp/postxconfig-NT-fv3lam.txt` file.  Unfortunately, outside of the variables that are currently plotted using the `plot_allvars` scripts, I'm not sure how to go about adding additional variables to the script.

For verification, since you are using data staged in EPIC's `/glade/work/epicufsrt/contrib/UFS_SRW_data/v2p2/input_model_data` directory, you should be able to point to the observations in `/glade/work/epicufsrt/contrib/UFS_SRW_data/v2p2/obs_data`.  These paths should already be defined in the `ush/machine/derecho.yaml` machine file.

To run verification in your experiment, please make sure that the following is in your `config.yaml` file:
```
rocoto:
  tasks:
    taskgroups: '{{ ["parm/wflow/prep.yaml", "parm/wflow/coldstart.yaml", "parm/wflow/post.yaml", "parm/wflow/verify_pre.yaml", "parm/wflow/verify_det.yaml"]|include }}'
```
If you also want to include plotting, replace the above with:
```
rocoto:
  tasks:
    taskgroups: '{{ ["parm/wflow/prep.yaml", "parm/wflow/coldstart.yaml", "parm/wflow/post.yaml", "parm/wflow/plot.yaml", "parm/wflow/verify_pre.yaml", "parm/wflow/verify_det.yaml"]|include }}'
```

**Reply by mkbenneh** [reply]
Going well, @MichaelLueken,

Is it possible to have access to an extended dataset for the verification task? I have other simulations covering periods not available in the path you provided.

I appreciate your help!

**Reply by MichaelLueken** [reply]
Hello, @mkbenneh,

EPIC only maintains data that is used for the workflow-end-to-end (WE2E) tests on Derecho.  I am unable to provide extended datasets for verification outside of the WE2E test staged data.  Unfortunately, you will need to stage any data that isn't currently available for your tests.

Mike Lueken

**Reply by mkbenneh** [reply]
Hello @MichaelLueken,

Got it. Thank you. I also noticed an issue with the SLP plot and wanted to find out what could be the possible issue here. 

Here is a path to my results and working directory - _"/glade/work/mbenneh/expt_dirs/term_project/2020062200/postprd"_

Regards,
Michael

![slp_regional_f036](https://github.com/user-attachments/assets/75a40a3f-95eb-4db7-8bf1-d38a5b097d12)

**Reply by MichaelLueken** [reply]
Hello @mkbenneh,

Unfortunately, I'm not sure what it happening in your SLP plots.  I've built a copy of the `release/public-v2.2.0` branch on Derecho and was able to replicate what you are encountering.  Changing CCPP physics schemes didn't change the behavior encountered, not did altering your SEUS_12km grid values.

I was wondering, is the grid shown in the plot the grid that you were expecting?  I saw in `ush/predef_grid_params.yaml` that the grid was supposed to include the SE US, Atlantic, and Sahara Desert.  The above grid doesn't include all of that domain.

I'll reach out to my colleagues and see if they might know what is causing this.

Mike

**Reply by mkbenneh** [reply]
Hello @MichaelLueken,

I am currently working on just the Southeastern US and will later extend my domain over the Atlantic and Sahara.

I will be looking forward to your response.

Kind regards, 
Michael

**Reply by mkbenneh** [reply]
Hello @MichaelLueken,

I have been able to resolve the issue from the image I sent earlier, attached here. However, I have a couple of questions regarding creating an ESGgrid, lambert comformal. I had to write a little script to do some math to calculate the WRTCMP_lon_lwr_left and WRTCMP_lat_lwr_left for the resolution I needed; otherwise, the forecast domain would be wrong. Which was pretty interesting. I would like to know the limit of WRTCMP_nx and WRTCMP_ny with respect to dy and dx.

![slp_regional_f012](https://github.com/user-attachments/assets/6f94a3d6-2de7-40cc-b065-559b4b73570d)

Kind regards,
Michael

**Reply by mkbenneh** [reply]
Hello @MichaelLueken,

Attached here is me trying to visualize my outputs, and I noticed something about my geoptential heights over certain areas like Carbo Verde, Santa Cruz de Tenerife, and the Islands around it. Same can be seen over areas over Cuba, Guatamala, Puerto Rico, etc. I would like to know the particular issue with this or the simulation.
![image](https://github.com/user-attachments/assets/431044b1-0180-4f15-a4d9-0ecbdd1b86ae)

**Comment by MichaelLueken** [comment, marked as answer]
Hello @mkbenneh,

I'm glad that you were able to figure out the issue with the plotting!

There shouldn't be any limits imposed on `WRTCMP_nx` and `WRTCMP_ny` with respect to dx and dy.

Due to differences in the map projection type (the native that is being used for computational purposes is the [extended schmidt gnomonic](https://dtcenter.org/sites/default/files/events/2020/2-purser-james.pdf) grid, while the write-component grid is lambert conformal), the grids don't correspond 1-to-1.  So, you should ensure that the compute grid (`ESGgrid_NX` and `ESGgrid_NY`) is a few gridpoints larger than your write-component grid (`WRTCMP_nx` and `WRTCMP_ny`), otherwise you will likely see undefined data in your model output.

For more details on suggestions for setting up a custom regional ESG grid, please see this [practical guide](https://dtcenter.org/ufs-short-range-weather-srw-practical-session-guide/session-2/3-defining-custom-regional-esg-grid) provided by DTC.

**Reply by mkbenneh** [reply]
Hello @MichaelLueken,

Alright, Thank you. I have further questions on the use of UPP and setting the wall time for longer simulations. Should I create a separate thread, or can I ask under here?

**Reply by MichaelLueken** [reply]
Hello @mkbenneh,

I would recommend creating a separate thread for questions related to UPP and setting the wall rime for longer simulations.

**Reply by mkbenneh** [reply]
@MichaelLueken 

Alright. In the config file in this directory, "_/work2/noaa/uswrp-hu/mbenneh/UFS-SRW/expt_dirs/dust_", I am unable to go beyond 8 hours. If I set the walltime to anything beyond 8 hours, it stops and says something about resource limits. I am not sure if there are ways to go around this problem.

**Reply by MichaelLueken** [reply]
@mkbenneh 

For Orion, the maximum limit for wall clock is 8 hours.  This is a hard limit imposed by MSU and there isn't anyway to get around the issue.
